### ISSUE TITLE
Ra reports CRC check failure in log segment

### ISSUE THREAD

**Issue by gibletto** [rabbitmq/ra#175]
We have a 5 node cluster using the RabbitMQ-HA Chart in Kubernetes on GKE.
We recently upgraded from 3.8.2 -> 3.8.3, and when the nodes came back up, some of the queues were in an error state., and could not be connected to. Both use OTP 22.3

I would say the upgrade was not graceful, while the nodes upgraded one at a time, the order is defined by K8S. We also had to use force_boot to restart the cluster

The following error pops up in logs for 3 of the Nodes.

````2020-04-02 12:32:43.440 [info] <0.31872.28> queue 'QueueName' in vhost '/': terminating with {case_clause,{'EXIT',ra_log_segment_crc_check_failure}} in state recover 
2020-04-02 12:32:43.529 [error] <0.31872.28> ** State machine '%2F_QueueName' terminating 
** Last event = {cast,go}
** When server state  = [{id,{'%2F_QueueNAme','rabbit@rabbitmq-rabbitmq-ha-discovery.svc.cluster.local'}},{opt,terminate},{raft_state,recover},{leader_last_seen,und 
** Reason for termination = error:{case_clause,{'EXIT',ra_log_segment_crc_check_failure}}
** Callback modules = [ra_server_proc]
** Callback mode = [state_functions,state_enter] 
** Stacktrace =  **  [{ra_log_reader,retry_read,4,[{file,"src/ra_log_reader.erl"},{line,144}]},{ra_log,take,3,[{file,"src/ra_log.erl"},{line,292}]},{ra_server,fetch_entries,3,[{file,"src/ra_server.erl"},{line,1935}]},{ra_server,apply_to,5,[{file,"src/ra_server.erl"},{line,1974}]},{ra_server,recover,1,[{file,"src/ra_server.erl"},{line,300}]},{ra_server_proc,recover,3,[{file,"src/ra_server_proc.erl"},{line,291}]},{gen_statem,loop_state_callback,11,[{file,"gen_statem.erl"},{line,1118}]},{proc_lib,init_p_do_apply,3,[{file,"proc_lib.erl"},{line,249}]}]  ````

.

**Comment by lukebakken** [comment]
@gibletto - I ran your question past @kjnilsson who wrote the code that deals with quorum queues. If you can provide an archive of RabbitMQ's data directory from a node that reports the `ra_log_segment_crc_check_failure` we can diagnose and fix this issue (if it is in fact a bug) quickly.

**Comment by gibletto** [comment]
I will wait for it to occur again. Unfortunately deleting was the only way forward. I has happened a few times in a few envs, I'll update when I get the data you need.

**Comment by michaelklishin** [comment]
If it does, please upload the Raft directory to this issue.

**Comment by lukebakken** [comment]
@michaelklishin I should have explained that I re-opened this after chatting with @kjnilsson . He can reproduce something similar and will continue looking at it tomorrow.

**Comment by gibletto** [comment]
I am pulling down a raft directory now.

Our queue has some quite large messages in it, and our WAL file is at 512mb also. 

It is possible to get a private place to send this on?

**Comment by michaelklishin** [comment]
`info@` RabbitMQ domain

**Comment by lukebakken** [comment]
@gibletto please compress the data, thanks! 😄

**Comment by kjnilsson** [comment]
This happening spontaneously without a reboot is very, very odd. It also happening for multiple queues at the same time after reboot is also very strange as the only way I can see this happening would be if the machine was rebooted before a segment was completely flushed (fsynced) to disk. The way the system works means that only a single segment per RabbitMQ node can be in this state at any one time which should limit this issue to a single queue member per node, not multiple.

Unless, of course, there is something else going on at the storage layer. I need as much information as possible about how this system is running, what storage system it uses and what guarantees it provides.

**Comment by kjnilsson** [comment]
@gibletto could it be possible that the errors have accumulated across multiple reboots and you only noticing when you've lost a quorum of nodes? So that perhaps 1 queue member corrupted per node reboot after all?

**Comment by kjnilsson** [comment]
@gibletto btw we don't need the WAL file if you want to reduce the size of what you are sending.

**Comment by gibletto** [comment]
It's very easy to replicate now on the existing cluster. Delete the queue, recreate and push our data to it. 
I will get you the segment files in a few hours, need to wake some people up!

**Comment by kjnilsson** [comment]
Deleting and re-creating a quorum queue creates a new underlying identity so I'm really not sure how that would cause it. Could I have the server logs for all nodes spanning the time from delete/create to the issue happening?

**Comment by kjnilsson** [comment]
At debug level ideally if possible.

**Comment by gibletto** [comment]
Setting up a test now

I noticed our segments are quite large.

```` 2F_MWGFPF55X1MJJX8$ ls -lh
total 4.4G
-rw-r--r-- 1 _apt 101 4.4G Apr  2 19:19 00000001.segment
-rw-r--r-- 1 _apt 101 2.2K Apr  2 19:20 config
drwxr-sr-x 2 _apt 101 4.0K Apr  2 17:48 snapshots

**Comment by kjnilsson** [comment]
segments contain a fixed number of records, you may benefit from lowering the `raft.segment_max_entries` config to say 4096.

**Comment by michaelklishin** [comment]
With messages in megabytes, I'd say it should be 256 or 512. I doubt there are pros to having gigantic segment files but there are cons as we see above.

**Comment by kjnilsson** [comment]
Yes for now you need to keep the segment size below ~4GB. Will look to address this asap

**Comment by kjnilsson** [comment]
do all the failed queues have segment sizes larger than ~4GB?

**Comment by kjnilsson** [comment]
Also @gibletto have these queues got a memory limit configured? https://www.rabbitmq.com/quorum-queues.html#memory-limit

**Comment by gibletto** [comment]
I believe that this particular issue is affecting a single queue. Most queues are smaller messages.

We have a 10mb limit applied to all queues as a policy.
max-in-memory-bytes:	10485760

I do believe we saw an error similar to "Channel Timeout" on the other queues in a different environment which trigger mass re-elections which never ended and ended with queues not starting. (Deleting fixed them).

It might have been on the back of this error. Our efforts are now on recreating this issue. I'm sorry for the delay, it's a production environment and I'm trying to either get a test on it, or rebuild a test environment.

**Comment by kjnilsson** [comment]
ok I'm pretty sure I know what the issue is with your large messages queue and I have a fix for it. I'm not sure why your queues with smaller messages fail in the same way so would be very interested in seeing the segment directory of such a queue if you can get it. it will be in the mnesia/quorum/<Node>/<DIR> where DIR is a partially random string with the vhost + queue name contributing to the first few characters.

**Comment by gibletto** [comment]
I'm seeing a different error for that other scenario

````[notice] <0.22433.194> queue 'QueueName' in vhost '/': candidate -> leader in term: 71456                                                                                                                                              
queue ''QueueName' ' in vhost '/': terminating with function_clause in state leader 
[error] <0.22433.194> ** State machine '%2F_QueueName ' terminating
Last event = {{timeout,tick},tick_timeout}                                                                                                                                                                                                                    ** When server state  = {leader,"ra_server_proc:format_status/2 crashed"}                                                                                                                                                                                     ** Reason for termination = error:function_clause                                                                                                                                                                                                             ** Callback modules = [ra_server_proc]                                                                                                                                                                                                                        ** Callback mode = [state_functions,state_enter]                                                                                                                                                                                                              ** Stacktrace =                                                                                                                                                                                                                                               **  [{rabbit_fifo,messages_ready,[{rabbit_fifo,{cfg,'%2F_QueueName',{resource,<<"/">>,queue,<<"QueueName">>},{64000,64000},{rabbit_quorum_queue,dead_letter_publish,[undefined,undefined,{resource,<<"/">>,queue,<<"QueueName">>}]},{rabbit_quorum_queue,become_leader,[{resource,<<"/">>,queue,<<"QueueName">>}]},undefined,undefined,competing,undefined,undefined,10485760},#{},undefined,3,{0,{[],[]}},2,#{},{rabbit_fifo_index,#{},undefined,undefined},{0,{[],[]}},#{{<<"amq.ctag-REoFQv8nO-mnav ││ 7yLngsqA">>,<9870.22937.3>} => {consumer,#{ack => true,args => [],prefetch => 16,username => <<"rabbituser">>},#{},0,16,0,simple_prefetch,auto,up},{<<"amq.ctag-Xxs7wv1sOAATlR18cE8jeQ">>,<9869.19056.0>} => {consumer,#{ack => true,args => [],prefetch => 16,username => <<"rabbituser">> },#{},0,16,0,simple_prefetch,auto,up},{<<"amq.ctag-ZNaH45MEYAExs385pRYLpQ">>,<0.2150.1>} => {consumer,#{ack => true,args => [],prefetch => 16,username => <<"rabbituser">>},#{},2,16,2,simple_prefetch,auto,up}},{[{<<"amq.ctag-REoFQv8nO-mnav7yLngsqA">>,<9870.22937.3>},{<<"amq.ctag-Xxs7 ││ wv1sOAATlR18cE8jeQ">>,<9869.19056.0>},{<<"amq.ctag-4lcHJNpUJ_Z8WV0MhVWlmw">>,<9869.17219.0>},{<<"amq.ctag-g1kdTIndeRY3MB9GhdxSGw">>,<9869.16201.0>},{<<"amq.ctag-daDOZR9dzKVbnc0AhmS2UQ">>,<9869.10632.0>}],[{<<"amq.ctag-ZNaH45MEYAExs385pRYLpQ">>,<0.2150.1>}]},{[],[]},0,0,[],0,0}],[{fi ││ le,"src/rabbit_fifo.erl"},{line,772}]},{rabbit_fifo,tick,2,[{file,"src/rabbit_fifo.erl"},{line,599}]},{ra_machine,tick,3,[{file,"src/ra_machine.erl"},{line,264}]},{ra_server_proc,leader,3,[{file,"src/ra_server_proc.erl"},{line,405}]},{gen_statem,loop_state_callback,11,[{file,"gen_st ││ atem.erl"},{line,1118}]},{proc_lib,init_p_do_apply,3,[{file,"proc_lib.erl"},{line,249}]}]                                                                                                                                                                                
[error] <0.22433.194> CRASH REPORT Process '%2F_QueueName' with 0 neighbours crashed with reason: no function clause matching rabbit_fifo:messages_ready({rabbit_fifo,{cfg,'%2F_QueueName',{resource,<<"/">>,queue,<<"QueueName">>},...},...}) line 772
[error] <0.1954.0> Supervisor {<0.1954.0>,ra_server_sup} had child '%2F_QueueName' started with ra_server_proc:start_link(#{await_condition_timeout => 30000,broadcast_time => 100,cluster_name => '%2F_QueueName',...}) at <0.22433.194> exit with reason no function clause matching rabbit_fifo:messages_ready({rabbit_fifo,{cfg,'%2F_QueueName',{resource,<<"/">>,queue,<<"QueueName">>},...},...}) line 772 in context child_terminated
[info] <0.23130.194> queue 'QueueName' in vhost '/': granting vote for {'%2F_QueueName','rabbit@rabbitmq-ha-0.rabbitmq-ha-discovery.svc.cluster.local'} with last indexterm {70058,71456} for term 71457 previous term was 71456 
[info] <0.23130.194> queue 'QueueName' in vhost '/': detected a new leader {'%2F_QueueName','rabbit@rabbitmq-ha-0.rabbitmq-rabbitmq-ha-discovery.svc.cluster.local'} in  term 71457                                                                           [info] <0.23130.194> queue 'QueueName' in vhost '/': Leader monitor down with {function_clause,[{rabbit_fifo,messages_ready,[{...}],[...]},{rabbit_fifo,tick,2,...},{ra_machine,tick,...},{ra_server_proc,...},{...}|...]}, setting election timeout          [info] <0.23130.194> queue 'QueueName' in vhost '/': granting vote for {'%2F_QueueName','rabbit@rabbitmq-ha-1.rabbitmq-ha-discovery.svc.cluster.local'} with last indexterm {70059,71457} for term 71458 previous term was 71458

**Comment by michaelklishin** [comment]
Let's focus on one issue at a time. @gibletto would you be able to test a Raft build by dropping it into `{installation_root}/plugins` in place of `ra-1.0.5.ez` or similarly named Ra code archive?

**Comment by gibletto** [comment]
I would be able to yes. The timeline on testing it is prob Monday though.

**Comment by michaelklishin** [comment]
Since the problem as currently understood is due to an integer range limitation that was put in place unintentionally (see #176), reducing message size so that segment files do not grow over 4 GiB would help in the meantime. Compression and/or putting large payloads in shared blob storage and passing around the key in messages are the two most common strategies for this.

**Comment by gibletto** [comment]
Oh absolutely, this pattern is due to be shot in the foot.

**Comment by gibletto** [comment]
The following values are applied to the values file for the chart to test this

````
extraConfig: |-
  raft.segment_max_entries = 512

**Comment by gibletto** [comment]
I can confirm, this setting worked without a hitch.

**Comment by michaelklishin** [comment]
We are QA'ing #176 and will upload a new Ra build once done.

**Comment by michaelklishin** [comment]
@gibletto can you please replace the `ra-1.0.8.ez` in the distribution with the (unarchived) file below and try without the `raft.segment_max_entries` config entry above? That would effectively test #176 in your environment.

[ra-1.1.0-eb5357c.ez.zip](https://github.com/rabbitmq/ra/files/4438319/ra-1.1.0-eb5357c.ez.zip)

**Comment by gibletto** [comment]
Is it expected that segments will be removed / compacted eventually?

**Comment by lukebakken** [comment]
@patrickmslatteryvt I recommend moving the original file to one with a different extension, in the same directory. Then, just move the `ra-1.1.0-eb5357c.ez.zip` file into `plugins/`. I don't believe it has to be named `ra-1.0.8.ez`, it just needs to have a `.ez` extension.

**Comment by michaelklishin** [comment]
The goal is to replace the Ra build you have in your distribution with this one. The runtime does not care what the file named but it has to have the `.ez` extension. GitHub does not allow for `.ez` file uploads so I had to compress it with `zip` (even though an `.ez` archive is already a zip file). Having two versions of the same library on code path is usually a terrible idea so yes, you must delete the old one.

Replacing the exact file can lead to confusion since you won't know for sure what version you are testing with. So I'd keep the name we have in the archive.

**Comment by michaelklishin** [comment]
Since we could reproduce the issue with Ra and added a test plus a segment file version migration, we consider this to be addressed and scheduled to ship with `3.8.4`. If there's more feedback on the Ra build above, let us know.
